spyder/api/widgets.py

```diff
@@ -0,0 +1,1478 @@
+            raise Exception('Wigdet with name "{}" already added. '
+                            'Current names are: {}'.format(
+                                name, list(self._widgets.keys())))
+
+        self._widgets[name] = widget
+        self._layout.insertWidget(0, widget)
+
+    def get_widget(self, name):
+        """
+        Return a widget by name.
+        """
+        if name in self._widgets:
+            return self._widgets[name]
+
+
+class SpyderMenu(QMenu):
+    """
+    A QMenu subclass to implement Spyder additional functionality.
+    """
+    MENUS = []
+
+    def __init__(self, name, parent=None, title=None):
+        self._name = name
+        self._sections = []
+        self._actions = []
+        self._ordered_actions = []
+        self._dirty = False
+
+        if title is None:
+            QMenu.__init__(self, parent)
+        else:
+            QMenu.__init__(self, title, parent)
+
+        self.MENUS.append((parent, name, title, self))
+        self.aboutToShow.connect(self._render)
+
+    def add_action(self, action, section=None, before=None):
+        """
+        Add action to a given menu section.
+        """
+        if before is None:
+            self._actions.append((section, action))
+        else:
+            new_actions = []
+            for sec, act in self._actions:
+                if act == before:
+                    new_actions.append((section, action))
+
+                new_actions.append((sec, act))
+
+            self._actions = new_actions
+
+        if section not in self._sections:
+            self._sections.append(section)
+
+        # Track state of menu to avoid re-rendering if menu has not changed
+        self._dirty = True
+        self._ordered_actions = []
+
+    def get_name(self):
+        """
+        Return name for menu.
+        """
+        return self._name
+
+    def get_sections(self):
+        """
+        Return a tuple of menu sections.
+        """
+        return tuple(self._sections)
+
+    def _render(self):
+        """
+        Create the menu prior to showing it. This takes into account sections
+        and location of menus. It also hides consecutive separators if found.
+        """
+        if self._dirty:
+            self.clear()
+            actions = []
+            for section in self._sections:
+                for (sec, action) in self._actions:
+                    if sec == section:
+                        actions.append(action)
+
+                actions.append(None)
+
+            # TODO: Remove any consecutive Nones
+            add_actions(self, actions)
+
+            self._ordered_actions = actions
+            self._dirty = False
+
+    @property
+    def name(self):
+        """
+        Return name for menu.
+        """
+        return self._name
+
+    @property
+    def sections(self):
+        """
+        Return a tuple of menu sections.
+        """
+        return tuple(self._sections)
+
+
+class MainWidgetMenu(SpyderMenu):
```

Comment:
Does this contain all the hacks we've added during the years to better support macOS? I don't see them in code below, so that would a terrible regression.

Reply:
Not sure what you mean. I saw some hacks on the `create_action` qthelpers which are still being used since I call that method.

For  add_actions I saw no hack.